utils/neuralnetwork.py

```diff
@@ -1,8 +1,53 @@
 import tensorflow as tf
 import numpy as np
+import tensorflow_graphics.math.interpolation as inter
 
 from custom_lbfgs import lbfgs, Struct
 
+class BSplineLayer(tf.keras.layers.Layer):
+    def __init__(self, layer_sizes, num_elem, degree):
+        super(BSplineLayer, self).__init__()
+        self.layer_sizes = layer_sizes
+        self.num_elem = num_elem
+        self.degree = degree
+
+    def get_spline(self, num_elem, x):
+        pos_x = x * num_elem
+        pos_spl = tf.cast(tf.math.floor(pos_x), dtype=tf.int32) + tf.constant(1) # let's get which element we're in...
+        pos_spl = tf.cond(pos_spl > num_elem, lambda: num_elem, lambda: pos_spl) # ...and not to get out of range (could be done better to avoid cond)
+        splines = inter.bspline._quadratic(pos_x%1.0) # comptue spline values (this impl support only [0,1] range)
+        return splines, pos_spl
+
+    def eval_point_2d(self, u, x, y, num_elem, degree): # compute `u` from 2D splines
+        splines_x, pos_x = self.get_spline(num_elem, x)
+        splines_y, pos_y = self.get_spline(num_elem, y)
+        X, Y = tf.meshgrid(splines_y, splines_x)
+        W = tf.slice(u, [pos_x-1, pos_y-1], [degree,degree]) # getting only those splines that are non-zero in this element
+        return tf.math.reduce_sum(W * X * Y) # Equivalent to: Sum_i_j { u_i_j * spline_x_i * spline_y_j }
+
+    def build(self, input_shape):
+        self.layers = [tf.keras.layers.Lambda(lambda X: X)] # Redundant I guess
+        for size in self.layer_sizes[1:-1]:
+            self.layers.append(tf.keras.layers.Dense(size, activation=tf.nn.tanh, kernel_initializer="glorot_normal"))
+        self.layers.append(tf.keras.layers.Dense(1, activation=tf.nn.tanh, kernel_initializer="glorot_normal"))
+        # BSpline indices can be done either by simple (i,j) encoding or one_hot_encoding
+        # Current implementation is one_hot_encoding, line below is for the reference of (i,j) encoding
+        # self.cofs = [[tf.math.floor(i / 7.0), i % 7.0] for i in range(49)]
+        input_shape = self.layer_sizes[0]
+        self.matrix_size = tf.math.sqrt(float(input_shape))
+        one_hot_encoded = []
+        for i in range(input_shape):
+            row = np.zeros((input_shape,))
+            row[i] = 1
+            one_hot_encoded.append(row)
+        self.cofs = tf.convert_to_tensor(one_hot_encoded, dtype=tf.float32) # cofs ->> coefficients' indices
+
+    def call(self, input):
+        output = self.cofs # first we need to ask our NN for coefficients for each 2D BSpline...
+        for layer in self.layers:
+            output = layer(output)
+        u = tf.reshape(output, [self.matrix_size,self.matrix_size]) * 3.0 # ...and reshape it. IMPORTANT! I think we need to multiply it to scale bc otherwise all coefficients will be almost 0 and with BSplines values result ends up as 0
```

Comment:
Would it make sense to work with log or something to play icon ce with small numbers?

Reply:
So with the `tanh` activation the output (coefficients) can be from -1 to 1. The spline basis functions are in range 0 to 1:
 ![](https://i.ibb.co/dB0X4p5/index.png)

So after multiplying basis and coefs I think it might be impossible to easily achieve greater values as final output like >8.0 here in `sin` initial condition.

That's the reason why I scaled output by multiplying by 3 (also tried different values), it gave bigger initial error but again was pulled close to zero as before.

I don't understand `play icon ce` - could you explain it to me?

Reply:
Sorry about this, autocorrect got me. I corrected it to "deal with small numbers"